### PR TITLE
Initial symcrypt + scossl commit

### PR DESCRIPTION
**Pull request by SpencerNofzinger**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x ] The toolchain has been rebuilt successfully (or no changes were made to it)
- [ ] The toolchain/worker package manifests are up-to-date
- [ ] Any updated packages successfully build (or no packages were changed)
- [ ] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [ ] All package sources are available
- [ ] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ ] All source files have up-to-date hashes in the `*.signatures.json` files
- [ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 
Adding SymCrypt and the Symcrypt Engine for OpenSSL as packages

###### Change Log  
- added SymCrypt package
- added SCOSSL package

###### Does this affect the toolchain?  
No

###### Test Methodology

- Pipeline build id: xxxx

### CODE REVIEW

**Comment by mlindgren** [comment]
Why have the libsymcrypt.so binary in the scossl spec? scossl should depend on symcrypt and link against the symcrypt binary in /usr/lib

File: SPECS/scossl/scossl.spec [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+%description
+The SymCrypt engine for OpenSSL (SCOSSL) allows the use of OpenSSL with SymCrypt as the provider for core cryptographic operations
+
+%prep
+
+%setup -n microsoft-SymCrypt-OpenSSL-ccdc1d9
+
+%build
+ls /usr/lib/
+cp /usr/lib/libsymcrypt.so ./
```
Comment (mlindgren) [inline SPECS/scossl/scossl.spec:26]: These lines shouldn't be necessary; we should just use the libsymcrypt.so in /usr/lib

File: SPECS/scossl/scossl.spec [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+%else
+cmake .. -DOPENSSL_ROOT_DIR=/usr/local/ssl -DCMAKE_TOOLCHAIN_FILE=../cmake-toolchain/LinuxUserMode-AMD64.cmake
+%endif
+cmake --build .
+
+%install
+mkdir -p %{buildroot}%{_libdir}
+mkdir -p %{buildroot}%{_includedir}
+install bin/SymCryptEngine/dynamic/libsymcryptengine.so %{buildroot}%{_libdir}/
+install SymCryptEngine/inc/scossl.h %{buildroot}%{_includedir}
```
Comment (mlindgren) [inline SPECS/scossl/scossl.spec:41]: SCOSSL has a `cmake install` target that you can use to install it. You should be able to just run `cmake install` here.

See [https://github.com/microsoft/SymCrypt-OpenSSL/blob/main/SymCryptEngine/dynamic/CMakeLists.txt](https://github.com/microsoft/SymCrypt-OpenSSL/blob/main/SymCryptEngine/dynamic/CMakeLists.txt)

File: SPECS/symcrypt/symcrypt.spec [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+%install
+mkdir -p %{buildroot}%{_libdir}
+mkdir -p %{buildroot}%{_includedir}
+%ifarch aarch64
+install bin/module/ARM64/LinuxUserMode/generic_linux/libsymcrypt.so %{buildroot}%{_libdir}/
+%else
+install bin/module/AMD64/LinuxUserMode/generic_linux/libsymcrypt.so %{buildroot}%{_libdir}/
+%endif
+install inc/* %{buildroot}%{_includedir}
```
Comment (mlindgren) [inline SPECS/symcrypt/symcrypt.spec:47]: We should add an install target to the SymCrypt CMake scripts as well, and then we can just call cmake install here too.

File: SPECS/scossl/scossl.signatures.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+    "Signatures": {
+     "scossl-0.1.tar.gz": "548bfd75a8a03ee9bfb75211d9968eb319bd7b79d03c92d5ebd5c12e5eae4143"
```
Comment (PawelWMS) [inline SPECS/scossl/scossl.signatures.json:3]: Can't comment on `libsymcrypt.so`, so adding this here: I think you've accidentally checked-in a library into the repo.:)

File: SPECS/scossl/scossl.spec [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+Summary:        The SymCrypt engine for OpenSSL (SCOSSL) allows the use of OpenSSL with SymCrypt as the provider for core cryptographic operations
+Name:           scossl
+Version:        0.1
+Release:        1%{?dist}
+License:        MIT License
+Vendor:         Microsoft Corporation
+Distribution:   Mariner
+Group:          System/Libraries
+URL:            https://github.com/microsoft/SymCrypt-OpenSSL
+Source0:        %{name}-%{version}.tar.gz
```
Comment (PawelWMS) [inline SPECS/scossl/scossl.spec:11]: Please use an URL for `Source0` pointing to the original source. This should match with the URL from the entry inside `cgmanifest.json` for this spec.

File: SPECS/symcrypt/symcrypt.spec [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+Summary:        A core cryptographic library written by Microsoft
+Name:           symcrypt
+Version:        100.20
+Release:        1%{?dist}
+License:        MIT License
+Vendor:         Microsoft Corporation
+Distribution:   Mariner
+Group:          System/Libraries
+URL:            https://github.com/microsoft/SymCrypt
+Source0:        %{name}-%{version}-with-submodules.tar.gz
```
Comment (PawelWMS) [inline SPECS/symcrypt/symcrypt.spec:11]: Same request for a URL.

File: SPECS/symcrypt/symcrypt.spec [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+BuildRequires:  python3-pyelftools
+
+%description
+A core cryptographic library written by Microsoft
+
+%prep
+%setup -n SymCrypt
+
+%patch0
+%patch1
```
Comment (PawelWMS) [inline SPECS/symcrypt/symcrypt.spec:27]: Suggestion: using [`%autosetup`](https://docs.fedoraproject.org/en-US/packaging-guidelines/#_autosetup) takes care of patching as well.

```suggestion
%autosetup -n SymCrypt
```

File: SPECS/symcrypt/symcrypt.spec [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+%ifarch aarch64
+install bin/module/ARM64/LinuxUserMode/generic_linux/libsymcrypt.so %{buildroot}%{_libdir}/
+%else
+install bin/module/AMD64/LinuxUserMode/generic_linux/libsymcrypt.so %{buildroot}%{_libdir}/
+%endif
+install inc/* %{buildroot}%{_includedir}
+
+%files
+%license LICENSE
+%{_libdir}/libsymcrypt.so
```
Comment (PawelWMS) [inline SPECS/symcrypt/symcrypt.spec:51]: Is this name correct? Commonly in Linux libraries are versioned, plain `A.so` files are just links and actual binaries have a form similar to `A.so.X.Y.Z` with another a link in between:

- `A.so` -> `A.so.X`;
- `A.so.X` -> `A.so.X.Y.Z`.

The `A.so` link is then part of the `devel` subpackage, which depends on the default package providing `A.X` and `A.X.Y.Z`. If the binary version doesn't have all `X.Y.Z` components (major, minor, and patch version numbers), then the middle link might fall out or the whole thing gets simplified and we're left with `A.so` -> `A.so.[some_other_versioning_scheme]`. Example in `ding-libs.spec`.

More info about conventions [here](https://medium.com/@manojddesilva/managing-linux-shared-libraries-be67ae14af2a), [here](https://unix.stackexchange.com/questions/475/how-do-so-shared-object-numbers-work), and [here](https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html).

**Comment by jslobodzian** [comment]
Closing this PR.  Symcrypt is now available on Mariner 2.0.

**Pull request closed without merging**